### PR TITLE
Support J operator in the type system

### PR DESCRIPTION
**Pull request by breuleux**
This adds the operators `J` and `Jinv`, the type `JTagged`, the MetaGraph `grad`, and type and shape inference support for all of them.

This is only the type system part, working implementation will be in the next PR.

### CODE REVIEW

File: myia/composite.py [hunk 1]
```diff
@@ -595,3 +597,79 @@ def _array_zero(xs):
...
+
+
+class GradOperation(MetaGraph):
+    """Implements the grad(f) operation.
+
+    grad(f)(x, ...) returns df(x, ...)/dx. Derivatives of other inputs are
+    thrown out.
+
+    TODO: This currently will not work on primitives, but it is an easy fix.
+    We just need to know how many parameters f takes.
```
Comment (abergeron) [inline myia/composite.py]: I feel that it's easy enough to wrap a primitive in a graph if you really need it that we don't need to implement that.

File: myia/infer/jinf.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
+"""Implements JInferrer."""
+
+from ..prim import ops as P
+from ..debug.label import short_relation_symbols as syms
```
Comment (abergeron) [inline myia/infer/jinf.py]: This is unecessary.

File: myia/composite.py (hunk 1 above)
Comment (breuleux) [reply myia/composite.py]: We'll see.

File: myia/infer/jinf.py (hunk 2 above)
Comment (breuleux) [reply myia/infer/jinf.py]: Helps a bit for debugging.

**Pull request merged**